nginx: make global redirect vhost option accept exceptions

vincentbernat opened the pull request:
###### Description of changes

By moving the return at the end of the host definition, one can provide exceptions by adding locations. Another solution would be to enclose the return into a location block (like what is done with forceSSL option).


<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

I didn't add a changelog entry because the changelog for the next NixOS version is not initialized yet. Hence keeping this PR as a draft for now.
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

vincentbernat commented:
Changelog entry added.

RaitoBezarius reviewed (approved):
Looks good to me, could you add a NixOS test for this?
I think we have no test for globalRedirect in `nixos/tests/nginx.nix`, but it would be a good time to add it. I can walk you through the process if you need help. :)

vincentbernat committed:
nginx: make global redirect vhost option accept exceptions

By moving the return into a location directive, one can provide
exceptions by adding locations. This is similar to what the forceSSL
option does.

Izorkin commented:
Working with this example locations?:
```
  locations."~ ^/example/(folder1|folder2)/" = {
```
or
```
    locations."~ ^/example/" = {
```

vincentbernat commented:
Regular expressions are used first and regular string locations are only used if none match, so, yes it should work.

mweinelt commented:
Can we inherit the new test into nginx's passthru.tests please?

vincentbernat commented:
@mweinelt updated!

vincentbernat commented:
Rebased.

mweinelt commented:
@ofborg test nginx-globalredirect
